sync-android/src/main/java/com/cloudant/sync/datastore/encryption/AndroidKeyProvider.java

```diff
@@ -0,0 +1,95 @@
+ * encrypted datastore.
+ * <p/>
+ * Given an user-provided password and an identifier, it generates a strong key and store it safely
+ * in the applications {@link android.content.SharedPreferences}, so the same key can be retrieved
+ * later provided that the user supplies the same password and id.
+ * <p/>
+ * The password is used to protect the key before saving it to the {@link android.content
+ * .SharedPreferences}. The identifier is an
+ * easy way to have more than one encryption key in the same app, the only condition is to provide
+ * different ids for each of them.
+ *
+ * @see KeyProvider
+ * @see KeyManager
+ * @see KeyStorage
+ */
+public class AndroidKeyProvider implements KeyProvider {
+    private static final Logger LOGGER = Logger.getLogger(AndroidKeyProvider.class
+            .getCanonicalName());
+    private String password;
+    private KeyManager manager;
+
+
+    /**
+     * Creates a {@link AndroidKeyProvider} containing a {@link KeyManager} associated with the
+     * provided identifier.
+     *
+     * @param context    The application's {@link Context}
+     * @param password   An user-provided password
+     * @param identifier The data saved in the {@link android.content.SharedPreferences} will be
+     *                   accessed with this identifier
+     * @see KeyManager
+     */
+    public AndroidKeyProvider(Context context, String password, String identifier) {
+        if (context != null && password != null && identifier != null && !password.equals("") &&
+                !identifier.equals("")) {
+            this.password = password;
+            KeyStorage storage = new KeyStorage(context, identifier);
+            this.manager = new KeyManager(storage);
+        } else {
+            LOGGER.severe("All parameters are mandatory");
+            throw new IllegalArgumentException("All parameters are mandatory");
+        }
+    }
+
+    /**
+     * Creates a {@link AndroidKeyProvider} containing the supplied {@link KeyManager}
+     *
+     * @param password An user-provided password
+     * @param manager  The {@link KeyManager} used to generate and persist the DPK.
+     * @see KeyManager
+     */
+    public AndroidKeyProvider(String password, KeyManager manager) {
+        if (password != null && manager != null && !password.equals("")) {
+            this.password = password;
+            this.manager = manager;
+        } else {
+            LOGGER.severe("All parameters are mandatory");
+            throw new IllegalArgumentException("All parameters are mandatory");
+        }
+    }
+
+    @Override
+    public EncryptionKey getEncryptionKey() {
+        EncryptionKey key = null;
+        if (this.manager.keyExists()) {
+            key = this.manager.loadKeyUsingPassword(this.password);
+        } else {
+            key = this.manager.generateAndSaveKeyProtectedByPassword(this.password);
+        }
+
+        return key;
+    }
```

Comment:
Changes requested in comment included in https://github.com/cloudant/sync-android/commit/dd276d263bfc9c5ddebb047cecd27230cb074f63